src/environment.rs

```diff
@@ -34,6 +43,87 @@ impl Environment {
+                                     0,
+                                     &mut desc_length_out as *mut raw::SQLSMALLINT,
+                                     string_buf,
+                                     0,
+                                     &mut attr_length_out as *mut raw::SQLSMALLINT);
+        }
+        loop {
+            match result {
+                raw::SQL_SUCCESS |
+                raw::SQL_SUCCESS_WITH_INFO => {
+                    count += 1;
+                    max_desc = std::cmp::max(max_desc, desc_length_out);
+                    max_attr = std::cmp::max(max_attr, attr_length_out);
+                }
+                raw::SQL_NO_DATA => break,
+                raw::SQL_ERROR => return Err(Error {}),
+                _ => unreachable!(),
+            }
+            unsafe {
+                result = raw::SQLDrivers(self.handle,
+                                         raw::SQL_FETCH_NEXT,
+                                         string_buf,
+                                         0,
+                                         &mut desc_length_out as *mut raw::SQLSMALLINT,
+                                         string_buf,
+                                         0,
+                                         &mut attr_length_out as *mut raw::SQLSMALLINT);
+            }
+        }
+
+        let mut driver_list = Vec::with_capacity(count);
+        loop {
+            let mut description_buffer: Vec<_> = (0..(max_desc + 1)).map(|_| 0u8).collect();
+            let mut attribute_buffer: Vec<_> = (0..(max_attr + 1)).map(|_| 0u8).collect();
+            unsafe {
+                result = raw::SQLDrivers(self.handle,
+                                         // Its ok to use fetch next here, since we know
+                                         // last state has been SQL_NO_DATA
+                                         raw::SQL_FETCH_NEXT,
+                                         &mut description_buffer[0] as *mut u8,
+                                         max_desc + 1,
+                                         &mut desc_length_out as *mut raw::SQLSMALLINT,
+                                         &mut attribute_buffer[0] as *mut u8,
+                                         max_attr + 1,
+                                         &mut attr_length_out as *mut raw::SQLSMALLINT);
+            }
+            match result {
+                raw::SQL_SUCCESS |
+                raw::SQL_SUCCESS_WITH_INFO => {
+                    description_buffer.resize(desc_length_out as usize, 0);
+                    driver_list.push(DriverInfo {
+                        description: String::from_utf8(description_buffer)
+                            .expect("String returned by Driver Manager should be utf8 encoded"),
+                        attributes: Self::parse_attributes(attribute_buffer),
+                    })
+                }
+                raw::SQL_ERROR => return Err(Error {}),
```

Comment:
I could not agree more. Yes we absolutely should! And in several other places also. In my next PR I'll be using the SQLDiagRec to get the Diagnostic records and transform them in an odbc::Error